src/sdk/pynni/nni/smac_tuner/smac_tuner.py

```diff
@@ -226,8 +229,16 @@ def generate_parameters(self, parameter_id, **kwargs):
         else:
             challengers = self.smbo_solver.nni_smac_request_challengers()
             for challenger in challengers:
+                if self.dedup:
+                    match = [v for k, v in self.total_data.items() \
+                             if v.get_dictionary() == challenger.get_dictionary()]
+                    if match:
+                        continue
                 self.total_data[parameter_id] = challenger
                 return self.param_postprocess(challenger.get_dictionary())
+            assert self.dedup is True
```

Comment:
I think we don't need the assert here

Reply:
i want it to have a quick fail

Reply:
check your logic and you could find it won't happen.

Reply:
what if `challengers` is empty?

Reply:
assert challengers not empty before the for loop will make it more straightforward

Reply:
https://github.com/automl/SMAC3/blob/81eba541e67c09aa307f2ba23d583f94c60b6125/smac/optimizer/ei_optimization.py#L699 this is the implementation of challengers, I am not sure whether it is good to use `len(challengers.challengers)`, seems very ugly.

Reply:
Got it, I agree that not to use challengers.chanllengers since it is intenal implementation. I am OK with this, or add another boolean variable to record and check whether challengers is empty.

Reply:
updated, thanks.